### DISCUSSION TITLE
`wrapped.__originalEntityData` serialization seems to vary, sometimes applying custom types, sometimes not

### DISCUSSION THREAD

**Discussion by leonardbinet** [mikro-orm/mikro-orm#6269] [Q&A]
I open a discussion, and not an issue because I've failed for now to replicate this bug in a simple reproduction repository ([wip here](https://github.com/leonardbinet/mikroorm-reproduction/pull/4/files)).

The issue I encounter is that since we migrated to v6.4.0 from v5.9.8, we see some unwanted sql updates that are triggered, while the entities didn't change. It is not critical since the operation is idempotent, and updates simply reset the right value for some relations.

It seems to be related to the serialization of foreign keys that have some custom type: in some specific deterministic circumstances that I still fail to understand, `helper.__originalEntityData` serializes some foreign keys with the custom type "db representation" (`string` in our case), whereas the entity's new state is always serialized with "js type" (`number` in our case). In `packages/core/src/unit-of-work/ChangeSetComputer.ts:112` this then results in an unwanted diff, that triggers an update.

Below some debugging screenshots on our codebase, that I failed for now to replicate in a simpler repo:
<img width="408" alt="Screenshot 2024-12-03 at 15 19 35" src="https://github.com/user-attachments/assets/08c0cac7-4481-4b73-8aed-089611a6a998">
<img width="274" alt="Screenshot 2024-12-03 at 15 21 47" src="https://github.com/user-attachments/assets/6c08cbf4-2baf-40c0-b479-590d99464403">
<img width="1107" alt="Screenshot 2024-12-03 at 15 21 58" src="https://github.com/user-attachments/assets/b522b70e-e608-4110-a6a3-b8aceb0645a8">

:
<img width="1671" alt="Screenshot 2024-12-03 at 15 17 36" src="https://github.com/user-attachments/assets/8d2588d9-6836-4a51-83ad-ee81d40cd000">

**Comment by leonardbinet** [comment]
another hint, still failing to replicate on a simple repo, but I'm getting closer: the order of fetch of entities matters.

I have three entities, `Analysis`, `AnalysisGroup` and `Trajectory`:

```
@Entity()
class AnalysisGroup {
  @PrimaryKey({type: IDStoredAsInteger, autoincrement: true})
  readonly id: string;
}

@Entity()
class Analysis {
  @PrimaryKey({type: IDStoredAsInteger, autoincrement: true})
  readonly id: string;

 @ManyToOne({ fieldName: 'analysisGroupId', deleteRule: 'NO ACTION', hidden: true })
  public analysisGroup: Ref<AnalysisGroup>
}

@Entity()
class Trajectory {
  @PrimaryKey({type: IDStoredAsInteger, autoincrement: true})
  readonly id: string;

@ManyToOne(() => Analysis, { deleteRule: 'CASCADE', fieldName: 'referenceAnalysisId', index: true })
  referenceAnalysis: Ref<Analysis>
}
```

depending on order of which is loaded first, orm will trigger or not the update:


OK: no change
```
const analysis = await this.analysisRepository.get({ id: analysisId })
const trajectory = await this.trajectoryRepository.get({ id: trajectoryId })
await this.entityManager.flush()
```

KO: triggers update of `analysis`, resetting `analysisGroupId` when trajectory is fetched beforehands:
```
const trajectory = await this.trajectoryRepository.get({ id: trajectoryId })
const analysis = await this.analysisRepository.get({ id: analysisId })
await this.entityManager.flush()
```

**Comment by B4nan** [comment]
From the diff it's clear that your problem is comparing `number` with `string`, so I guess your custom type is a wrapper around bigint? If so, make sure you extend the `BigIntType` there. Either way, let's start with you presenting the custom type implementation. Or at least make sure your `convertToDatabaseValue` is normalizing the value to string:

https://github.com/mikro-orm/mikro-orm/blob/ee24f1f0f017d8e7b4544f6cb1a39eab29dfb413/packages/core/src/types/BigIntType.ts#L20

Issues like this often originates from loading nested relations via joined strategy.

**Reply by B4nan** [reply]
Looking at [the type implementation here](https://github.com/leonardbinet/mikroorm-reproduction/pull/4/files#diff-cbdbf107a7c1ce7535e462b2471bb3e8a082f4ad06ab3eed8bc7b4c486d17d05R16), you might be able to get around this by implementing the `compareValues` method in there, and normalizing the values to `string` or `number` explicitly just for the comparison. 

It still sounds like something to fix on our end, so I would still appreciate a repro, but at least this way you should be able to get around it in your app.

**Reply by leonardbinet** [reply]
exactly, thank you very much indeed it works by overriding the `compareValues` method: here is the wip-repro, which mimics our codebase, but still fails to reproduce the issue

https://github.com/leonardbinet/mikroorm-reproduction/pull/5/files
